couchbase_test.go

```diff
@@ -0,0 +1,758 @@
+			},
+		},
+	}
+	resource, err := pool.RunWithOptions(ro)
+	if err != nil {
+		t.Fatalf("Could not start local couchbase docker container: %s", err)
+	}
+
+	cleanup = func() {
+		docker.CleanupResource(t, pool, resource)
+	}
+
+	address := fmt.Sprintf("http://127.0.0.1:8091/")
+	if err = pool.Retry(func() error {
+		t.Log("Waiting for the database to start...")
+		resp, err := http.Get(address)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
+			return fmt.Errorf("Got a %d status code from couchbase's Web Console", resp.StatusCode)
+		}
+		return nil
+	}); err != nil {
+		t.Fatalf("Could not connect to couchbase: %s", err)
+	}
+
+	containerInitialized = true
+
+	return cleanup, "0.0.0.0", 8091
+}
+
+func TestDriver(t *testing.T) {
+	// Spin up couchbase
+	cleanup, address, port := prepareCouchbaseTestContainer(t)
+
+	defer cleanup()
+
+	err := createUser(address, port, adminUsername, adminPassword, "rotate-root", "rotate-rootpassword",
+		"rotate root user", "admin")
+	if err != nil {
+		t.Fatalf("Failed to create rotate-root test user: %s", err)
+	}
+	err = createUser(address, port, adminUsername, adminPassword, "vault-edu", "password",
+		"Vault education user", "admin")
+	if err != nil {
+		t.Fatalf("Failed to create vault-edu test user: %s", err)
+	}
+
+	t.Run("Version", func(t *testing.T) { testGetCouchbaseVersion(t, address) })
+
+	if !pre6dot5 {
+		err = createGroup(address, port, adminUsername, adminPassword, "g1", "replication_admin")
+		if err != nil {
+			t.Fatalf("Failed to create group g1: %s", err)
+		}
+		err = createGroup(address, port, adminUsername, adminPassword, "g2", "query_external_access")
+		if err != nil {
+			t.Fatalf("Failed to create group g1: %s", err)
+		}
+	} else {
+		t.Log("Skipping group creation as the Couchbase DB does not support groups")
+	}
+
+	t.Run("Init", func(t *testing.T) { testCouchbaseDBInitialize_TLS(t, address, port) })
+	t.Run("Init", func(t *testing.T) { testCouchbaseDBInitialize_NoTLS(t, address, port) })
+	t.Run("Init", func(t *testing.T) { testCouchbaseDBInitialize_Pre6dot5TLS(t, address, port) })
+	t.Run("Init", func(t *testing.T) { testCouchbaseDBInitialize_Pre6dot5NoTLS(t, address, port) })
+
+	/* Need a pause as sometimes the travel-sample bucket is not ready and you get strange errors like this...
+	   err: {"errors":{"roles":"Cannot assign roles to user because the following roles are unknown, malformed or role
+	       parameters are undefined: [bucket_admin[travel-sample]]"}}
+	   Can use http://Administrator:password@localhost:8091/sampleBuckets to see if the installed element is true before
+	   proceeding.
+	   [{"name":"beer-sample","installed":false,"quotaNeeded":104857600},
+	    {"name":"gamesim-sample","installed":false,"quotaNeeded":104857600},
+	    {"name":"travel-sample","installed":false,"quotaNeeded":104857600}] */
+
+	time.Sleep(10 * time.Second) // need a pause as sometimes the travel-sample bucket is not ready.
```

Comment:
It will be quite time consuming to write a test that looks for this JSON object to change from false to true {"name":"travel-sample","installed":false,"quotaNeeded":104857600} and real fiddly to test as on my local machine, it did not fail, only on the slower Circleci VM. The cheap sleep seems to work, in my humble opinion.

Reply:
Changed my mind and added a back off function to check to see if the bucket is installed before proceeding with the tests that need a bucket ready.